compiler/src/iree/compiler/GlobalOptimization/DataLayoutUtils.h

```diff
@@ -0,0 +1,179 @@
+    return tf;
+  }
+
+private:
+  /// Transform the layout as if propagating through an operation, from the
+  /// `currentValue` to `newValue`, and return the new layout.
+  bool transform(Operation *op, Value currentValue, Value newValue);
+
+  /// The original type corresponding to the source of this layout.
+  ShapedType originalType;
+  /// The type of the layout source after the transformation is applied.
+  ShapedType transformedType;
+  /// Transformation metadate from `originalType`->`transformedType, represented
+  /// as pack metadata (innerDimsPos, innerTileSizes, outerDimsPerm) for now.
+  SmallVector<int64_t> innerDimsPos;
+  SmallVector<int64_t> innerTileSizes;
+  SmallVector<int64_t> outerDimsPerm;
+  /// Optional padding value for packed layouts.
+  std::optional<TypedAttr> constantPadValue = std::nullopt;
+  /// Indices in the `originalType` corresponding to each index of the value
+  /// associated with this DataLayoutTransformation.
+  SmallVector<int64_t> correspondingTransformedIndices;
+};
+
+//===----------------------------------------------------------------------===//
+// Analysis helpers
+//===----------------------------------------------------------------------===//
+
+/// Analyze the producer and users of this value, and return the node type for
+/// the given value.
+DataLayoutNodeType getNodeTypeForValue(Value value);
+
+/// If the node is a terminal node (e.g., the source of a layout, like a value
+/// next to a global load or store), then return the layoutIDs corresponding to
+/// that terminal node.
+SmallVector<StringRef> getTerminalNodeIDs(Value value);
+
+//===----------------------------------------------------------------------===//
+// Pass helpers
+//===----------------------------------------------------------------------===//
+
+/// Rewrite a Util::GlobalOp into the new layout indicated by the given
+/// DataLayoutTransformation, and rewrite all GlobalLoadOps or GlobalStoreOps
+/// of the `global`, passed in `edgeNodes` as the results or inputs to the
+/// load/store ops. Also, create an initializer to fill the new packed global
+/// with the padding value of the pack in the `transform`.
+LogicalResult
+transformGlobalsToNewLayout(IRRewriter &rewriter, SmallVector<Value> edgeNodes,
+                            DataLayoutTransformation *transform,
+                            const Explorer::GlobalInfo *globalInfo,
+                            SymbolTable moduleSymbols);
+
+//===----------------------------------------------------------------------===//
+// Attribute helpers
+//===----------------------------------------------------------------------===//
+
+/// Padding values can get in the way of unpack(pack(x)) foldings, so this sets
+/// a `__foldable_pack_unpack__` attribute, which indicates that the op came
+/// from a given data layout, and can be folded with the inverse of the op as
+/// long as it also has the same attribute.
+void setFoldablePackUnPackAttribute(Operation *op);
```

Comment:
Such attributes shouldnt leak pass boundaries. So it is better to avoid having them in a header file, but rather make them local to a pass implementation.

Reply:
This attribute is used in the util functions and also the pass patterns, and I wanted to keep the util stuff out of the pass implementation. Would it be fine to just rename the util files to something like `PropagateDataLayoutUtils` so it is more clear that they are specific to this pass?

Reply:
Maybe create a folder for pack propagation and move everything into that.